File: leetcode/longest-substring-without-repeating-characters.java
//https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int max = 0;
        int currLen = 0;
        char[] input = s.toCharArray();
        
        for(int i = 0; i<input.length; i++){
            Integer lastSeen = map.get(input[i]);
            if(lastSeen != null && i-lastSeen <= currLen){
                max = Math.max(max, currLen);
                currLen = i-lastSeen;
            }else{
                currLen++;
            }
    
            map.put(input[i], i);
        }
        
        return Math.max(max, currLen);
    }
}

File: leetcode/longest_substring_without_repeating_characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen: dict[str, int] = {}
        longest: int = 0
        start: int = 0

        for index, char in enumerate(s):
            seen = last_seen.get(char)
            if seen is not None and seen >= start:
                longest = max(longest, index - start)
                start = seen + 1
            last_seen[char] = index

        return max(longest, len(s) - start)
